### ISSUE TITLE
K-complex detection

### ISSUE THREAD

**Issue by remrama** [raphaelvallat/yasa#184]
Hi Raphael,
Sorry to disturb. One more stupid question. Is yasa also supportable in detecting k-complexes and vertex sharp waves?
Or I should switch to other tools to achieve this aim?

Thank you very much!
Zixiao

_Originally posted by @zixiao-yin in https://github.com/raphaelvallat/yasa/issues/40#issuecomment-937435148_

**Comment by remrama** [comment]
> Hi @zixiao-yin,
> 
> You should be able to detect k-complexes with the yasa.sw_detect function. Make sure that you limit the detection to N2 sleep. Detection of sharp waves is not implemented.
> 
> Thanks! Raphael

_Originally posted by @raphaelvallat in https://github.com/raphaelvallat/yasa/issues/40#issuecomment-938110694_

**Comment by remrama** [comment]
Thanks for the great package :) 

Reviving the question on K-complexes, are there consensus to say that K-complexes are simply N2 slow oscillations? while I do find some papers supporting this point, there are many others saying otherwise. There are also definitions for the KC waveform to be of complete reverse shape i.e. strong surface positive followed by strong surface negative) I have attached some screengrab from papers 

![GetImage (1)](https://github.com/user-attachments/assets/0932cfd7-5243-4332-8459-72f5ce0efa4a)

<img width="375" alt="image" src="https://github.com/user-attachments/assets/0624cf05-53fe-4a7d-b973-04d983e7360e">

[Halasz1998_Hierarchy_of_micro_arousals_and_the_microstructure_of_sleep.pdf](https://pubmed.ncbi.nlm.nih.gov/9894227/)

<img width="291" alt="GetImage" src="https://github.com/user-attachments/assets/ca5030af-311e-4aa2-9e13-77da07f37f7c">

https://www.scientificbulletin.upb.ro/rev_docs_arhiva/full782_186975.pdf

of course they could easily be montage choices etc, but would be great if the detection could also take the reverse shaped waveform into accounrt. Happy to contribute of course

_Originally posted by @Jhanyi in https://github.com/raphaelvallat/yasa/issues/40#issuecomment-2526019034_

**Comment by remrama** [comment]
Sorry for the barrage here, but I wanted to move the conversation about K-complexes over here from Issue #40 because there were two separate conversations going on their simultaneously.

Hi @Jhanyi and sorry for moving your post, but maybe we can address it here. I can't speak to consensus, but if you were wanting to try and detect the "reverse" shaped waveform, maybe you could hack it through YASA by setting the negative-peak and positive-peak amplitudes to negative values? I don't know if YASA would accept this (I haven't worked on the `sw_detection`), but since it seems to look for the negative-peak first and positive-peak second, maybe flipping the signs of those would just do the opposite. Not sure, but let me know if you've already played around with the `sw_detection` parameters.

**Comment by Jhanyi** [comment]
@remrama no problem, thanks for creating a separate issue. So I did have a play around switching the postive and negative peak, and it does seem to work well on my suspected KCs with that reverse waveform.  I also thought of just inverting the data array (simply data*-1) and apply yasa.sw_detect on this. Both seem to agree quite well on N2, but poorly on N3, which is expected as it should follow the SW morphology. I also loosened both parameters a bit 
N2
![KC_rev_detect_eg2](https://github.com/user-attachments/assets/f5209b93-916f-4aae-bf1e-884be294096c)
![KC_rev_detect_eg5](https://github.com/user-attachments/assets/3034a96a-fad4-4103-821f-9d2ab5d6b4d2)

N2 but poor agreement and some misses
![KC_rev_detect_eg1](https://github.com/user-attachments/assets/560186f2-fdd1-4e1b-952a-a69f4ebbbbff)

N3 - not great, but expected?
![KC_rev_detect_eg7](https://github.com/user-attachments/assets/db503c85-9bb4-4b57-bca0-70c31cde6fc0)
![KC_rev_detect_eg8](https://github.com/user-attachments/assets/55e45c73-4a51-4d98-82e9-416872733e12)

**Comment by Jhanyi** [comment]
> So if I understand correctly, the plots titled `reverse peak sequence on raw data` are also using sw_detect, just by inverting the detection parameters as opposed to the data (please correct me if I'm wrong). 
yep :) so finding positive peak first, then the following negative peak, and the data remains untouched



> but my intuition would be to prefer the parameter-manipulation bc it only impacts how the algorithm works and doesn't influence the raw data in any way
Agreed - so I guess this shall just be a temporary solution for now
